scraping/src/commonJvmAndroid/kotlin/cz/lastaapps/scraping/AgataWallet.kt

```diff
@@ -0,0 +1,94 @@
+package cz.lastaapps.scraping
+
+import io.ktor.client.HttpClient
+import io.ktor.client.plugins.defaultRequest
+import io.ktor.client.request.forms.FormDataContent
+import io.ktor.client.request.get
+import io.ktor.client.request.header
+import io.ktor.client.request.post
+import io.ktor.client.request.setBody
+import io.ktor.client.statement.bodyAsText
+import io.ktor.http.Parameters
+import it.skrape.core.htmlDocument
+import java.net.URLDecoder
+
+actual class AgataWallet {
+    private val client: HttpClient = HttpClient {
+        // Disable redirects, because when I was testing it with Python I got into loop sometimes
+        // Also it has to extract some cookies from some of the requests, so redirects manually handled
+        followRedirects = false
+
+        defaultRequest {
+            header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36")
+        }
+    }
+
+    /// Get balance from Agata
+    actual suspend fun getBalance(username: String, password: String): Float {
+        // Go to the auth provider
+        var response = client.get("https://agata.suz.cvut.cz/secure/index.php")
+        var url = response.headers["Location"]
+        response = client.get(url!!)
+
+        // Get new url params
+        var body = response.bodyAsText()
+        val returnUrl = Regex("var returnURL = \"(.+?)\"").find(body)?.groups?.get(1)?.value
+        val otherParams = Regex("var otherParams = \"(.+?)\"").find(body)?.groups?.get(1)?.value
+        url = "${returnUrl}&entityID=https://idp2.civ.cvut.cz/idp/shibboleth${otherParams}"
+
+        // Get to SSO
+        response = client.get(url)
+        response = client.get(response.headers["Location"]!!)
+
+        // Extract JSESSIONID cookie
+        val jsessionid = response.headers["Set-Cookie"]!!.split(";")[0].split("=")[1]
+
+        // Resolve SSO
+        url = "https://idp2.civ.cvut.cz${response.headers["Location"]}"
+        client.get(url)
+        response = client.post(url) {
+            setBody(FormDataContent(Parameters.build {
+                append("j_username", username)
+                append("j_password", password)
+                append("_eventId_proceed", "")
+            }))
+            header("Cookie", "JSESSIONID=${jsessionid}")
+            header("Referer", url)
+            header("Content-Type", "application/x-www-form-urlencoded")
+        }
+
+        // Extract response codes from html
+        var html = htmlDocument(response.bodyAsText())
+        var relayState: String? = null
+        var samlResponse: String? = null
+        html.findAll("input").forEach {
```

Comment:
Jak těžký by bylo vytáhnout tu hodnotu pomocí regexu? Protože téhle knihovny se chci v blízké budoucnosti zbavovat...

Reply:
https://stackoverflow.com/a/1732454/8220686

Reply:
Jakoze chapu to, ale je mi to celkem dost proti srsti. HTML (XML) parser vzdy bude spolehlivejsi (np.: poradi attributu, mezery a " ').
Taky imo kod z HTML parserem a CSS selectory je o dost kratsi, citelnejsi a maintainable nez nejake `class=\"[^\"]+\"`, a + to bude taky delsi protoze je potreba rucniho url decode, a 3194819491 null checku.

Ale nemam problem to prepsat ig, jen mi to pripada jako downgrade.

Reply:
Ano, je to downgrade, ale neni to platform dependent a nenašel jsem mupltiplatformní scraper. Jinak samo souhlas.

Reply:
Ja myslel ze ten scrapper je cisty Kotlin/Java, takze bezi v JVM = platform independent. Ale jestli neni tak radeji ti to rewritnu v Rustu a udelam CI na vsechny platformy nez abych parsoval HTML regexem.

Reply:
Okej, spíš target independent než platform.

A to by mohla bejt dost sranda udělat to intercompatibilní. Ještě ale teda pokouknu, už je to docela dlouho co jsem to hledal. Ozvu se pak

Reply:
Dobro, kompromis. Našel jsem knihovnu, co by snad mohla být dostatečně ergonomická. Hoď tam ale prosím ten regex do komentáře jako odstrašení pro lidi, co nechtějí strávit 5 min hledáním na netu jako já :sweat_smile:.

https://github.com/MohamedRejeb/Ksoup

Reply:
👍 kouknu se na to vecer po 5 pivech

Reply:
Ted jsem se koukal na KSoup, a vubec se mi nelibi to API / ono by to celkem zkomplikovalo to parsovani (a possiby spomalilo). Hledal jsem jine knihovny, ale nic jsem nenasel. Zatim to necham z tim skrape.it, protoze se mi libi ta API, a nevidim proc by to nebylo cross platform. 

<sub><i>Pri nejhorsim tady pridam RIIR commit </i> </sub>

Reply:
Okej, buďiž.